enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType
+
+	// Action can be Allow or Deny.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Allow";"Deny"
+	Action IngressNodeFirewallActionType `json:"action" protobuf:"bytes,4,opt,name=action,casttype=IngressNodeFirewallActionType"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type IngressNodeFirewallRuleProtocolType string
+
+const (
+	// ProtocolTypeICMP refers to the ICMP protocol
+	ProtocolTypeICMP ProtocolType = "ICMP"
+
+	// ProtocolTypeICMPv6 refers to the ICMPv6 protocol
+	ProtocolTypeICMPv6 ProtocolType = "ICMPv6"
+
+	// ProtocolTypeTCP refers to the TCP protocol, for either IPv4 or IPv6
+	ProtocolTypeTCP ProtocolType = "TCP"
+
+	// ProtocolTypeUDP refers to the UDP protocol, for either IPv4 or IPv6
+	ProtocolTypeUDP ProtocolType = "UDP"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol, for either IPv4 or IPv6
+	ProtocolTypeSCTP ProtocolType = "SCTP"
+)
+
+// IngressNodeFirewallActionType indicates whether an IngressNodeFirewallRule allows or denies traffic
+// +kubebuilder:validation:Pattern=`^Allow|Deny$`
+type IngressNodeFirewallActionType string
+
+const (
+	IngressNodeFirewallAllow IngressNodeFirewallActionType = "Allow"
+	DIngressNodeFirewallDeny IngressNodeFirewallActionType = "Deny"
+)
+
+// IngressNodeFirewallRules define ingress node firewall rule
+type IngressNodeFirewallRules struct {
+// FromCIDRS is A list of CIDR from which we apply node firewall rule
+FromCIDRs []IngressNodeFirewallFromCIDR `json:"fromCIDRs" protobuf:"bytes,1,opt,name=fromCIDRs"`
+// FirewallProtocolRules is A list of per protocol ingress node firewall rules
+FirewallProtocolRules []IngressNodeFirewallProtocolRule `json:"firewallProtocolRules" protobuf:"bytes,2,opt,names=firewallProtocolRules"`
+}
+
+// NodeIngressFirewallConfigConfigSpec consists of a node's ingress firewall rules
+// +k8s:openapi-gen=true
+type NodeIngressFirewallConfigSpec struct {
+	// A list of ingress firewall policy rules.
+	// empty list indicates no ingress firewall i.e allow all incoming traffic.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Ingress []IngressNodeFirewallRules `json:"ingress" protobuf:"bytes,2,opt,name=ingress"`
+}
+
+
+// NodeEndPointConfigSpec consists of list of interfaces where ingress node firewall policy will be attached to
+// +k8s:openapi-gen=true
+type NodeEndPointConfigSpec struct {
+	// A list of interfaces where the ingress firewall policy will be applied on.
+	// empty list indicates the firewall policy applied on all interfaces
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:UniqueItems:=true
+	// +optional
+    // +nullable
+	Interface *[]string `json:"interface" protobuf:"bytes,1,opt,name=interface"`
```

Comment:
Does this need to be a pointer type?  what is the difference between nil and empty?

Reply:
in case u don't specific any interfaces at all in the spec so we will have null pointer while empty list still hold some memory but len is 0